### ISSUE TITLE
Java14 Full Records Support Check Validation: WhitespaceAroundCheck

### ISSUE THREAD

**Issue by nmancus1** [checkstyle/checkstyle#8528]
Child of #8452 
Check documentation: https://checkstyle.sourceforge.io/config_whitespace.html#WhitespaceAround

From check documentation:
_Checks that a token is surrounded by whitespace. Empty constructor, method, class, enum, interface, loop bodies (blocks), lambdas of the form..._

```
➜  full-record-grammar /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 TestClass.java
Note: TestClass.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
➜  full-record-grammar cat config.xml    
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
      <module name="WhitespaceAround"/>
  </module>
</module>
➜  full-record-grammar cat TestClass.java
class MyClass {} // violation 2x
interface Foo {} // violation 2x

//simple record def
record MyRecord() {} // violation 2x

//simple record def
record MyRecord1() { } // ok

//nested constructs
record MyRecord2() {
    class MyClass {} // violation 2x
    interface Foo {} // violation 2x
    record MyRecord() {} // violation 2x
}

//method
record MyRecord3() {
    void method(){ // violation
        final int a = 1;
        int b= 1; // violation
        b=1; // violation 2x
    }

}

//ctor
record MyRecord4() {
    public MyRecord4() {
        final int a = 1;
        int b= 1; // violation
        b=1; // violation 2x
    }
}

//compact ctor
record MyRecord5() {
    public MyRecord5 {
        final int a = 1;
        int b= 1; // violation
        b=1; // violation 2x
    }
}

//static fields
record MyRecord6() {
    static final int a = 1;
    static int b= 1; // violation
}
➜  full-record-grammar java $RUN_LOCALE -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.35-SNAPSHOT-all.jar -c config.xml TestClass.java
Starting audit...
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:1:15: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:1:16: '}' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:2:15: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:2:16: '}' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:5:19: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:5:20: '}' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:12:19: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:12:20: '}' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:13:19: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:13:20: '}' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:14:23: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:14:24: '}' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:19:18: '{' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:21:14: '=' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:22:10: '=' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:22:10: '=' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:31:14: '=' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:32:10: '=' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:32:10: '=' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:40:14: '=' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:41:10: '=' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:41:10: '=' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:48:17: '=' is not preceded with whitespace. [WhitespaceAround]
Audit done.
Checkstyle ends with 23 errors.

```
This check works as intended with the new records syntax, so I think we can close this issue.

**Comment by pbludov** [comment]
This check should handle compact record ctors too:

```xml
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="WhitespaceAround">
            <property name="allowEmptyConstructors" value="true"/>
        </module>
    </module>
</module>
```

```java
class TestClass {
    public TestClass() {} // ok
}

record TestRecord() {
    public TestRecord {} // unexpected violation x2
}
```

The method `WhitespaceAroundCheck.isEmptyCtorBlock` should be ware of `TokenTypes.COMPACT_CTOR_DEF`.
Sure, this looks ridicules, empty compact ctor makes no sense, but who knows? It may be necessary for an annotation or something like that.

**Comment by nmancus1** [comment]
@pbludov This should be a violation too, right?

```
record TestRecord () {
              // ^
    public TestRecord {} 
}
```

Also, what about allowing violation of empty component list, `allowEmptyComponentList` option?

**Comment by pbludov** [comment]
> @pbludov This should be a violation too, right?

Yes, it should be more like a ctor declaration.

```java
class TestClass {
  TestClass () { // if there is a violation...
  }
}
class TestRecord () { // ...then there must be the same violation here
}

```
> Also, what about allowing violation of empty component list, `allowEmptyComponentList` option?
Not sure. There is already `allowEmptyTypes`. Also `allowEmptyConstructors` for compact ctors.

**Comment by nmancus1** [comment]
Actually, there is no violation  for this:
```
    class MyClass2 {
        MyClass2 () {
             // ^ not a violation
        }
    }
```

**Comment by pbludov** [comment]
> Actually, there is no violation for this:

Both `MyClass2 ()` and `MyClass2()` are valid? If so, there is no need to report a violation for records here.

**Comment by nmancus1** [comment]
I'm adding test cases to show this to the inputs.

**Comment by strkkk** [comment]
Fix is merged
